### PR TITLE
Unified training loop + BERT-style MAE + scaling-study configs

### PR DESCRIPTION
**Pull request by ksd3**
## Unified training loop + BERT-style MAE + scaling-study configs

Consolidates the AstroPT training-loop work into **one configurable script** plus the scaling-study run matrix. **Supersedes #63, #64, #65, #66** (all folded in here).

### Model (single image modality)
- **BERT-style MAE** objective: mask tokens flow through the full bidirectional encoder, masked patches are reconstructed (`objective="mae"`, `attn_type="full"`); the AR path is unchanged.
- **aim/affine tokeniser** selectable via config (wired into `model_args`).

### Unified `scripts/train.py` (replaces train.py + train_mae.py)
- `objective` (`"ar"`/`"mae"`) is a config flag driving `process_modes`, `model_args`, and a branched `validate()`.
- DDP **stream sharding** (`split_dataset_by_node` + buffered shuffle).
- **step checkpointing** (`num_checkpoints`, `checkpoint_schedule` = log/even/manual; first & last always pinned).
- tokeniser + **modalities built from config** (`patch_size`).
- `train_mae.py` deleted. The LLM / AION / multimodal scripts stay separate (different backbones, out of scope).

### 12 study configs (`config/study/`)
`{AR, MAE} × {aim, affine} × {1M, 21M, 100M}`, 64 log-spaced checkpoints, Pythia-style per-size LR (1M/21M → 1e-3, 100M → 6e-4). See `config/study/README.md`.

### Testing
Ran the unified script **end-to-end on synthetic data for both objectives** — train, eval, reconstruction previews, scheduled checkpoints, and reload via `load_astropt` (objective/tokeniser preserved). Model-level: mask ratio, gradients to the mask token, masked-only loss, AR path unchanged.

### CODE REVIEW

**Comment by Smith42** [comment]
Can we do 64 checkpoints per run?

Also, I am not sure if we want to change the LR or not with model size. EleutherAI do this in Pythia. May be a question for your AI assistant as they will have more context of the SOTA?

**Comment by ksd3** [comment]
> Can we do 64 checkpoints per run?
> 
> Also, I am not sure if we want to change the LR or not with model size. EleutherAI do this in Pythia. May be a question for your AI assistant as they will have more context of the SOTA?

Let me check that too.

File: scripts/train.py
Comment (Smith42) [inline scripts/train.py:1]: a lot of useful comments are deleted here, can you have a manual check and keep the useful ones please?

File: scripts/train.py [hunk 1]
```diff
@@ -643,73 +538,47 @@ def save_checkpoint(filename):
...
+            )
             dts = []
 
         iter_num += 1
         local_iter_num += 1
 
-        # termination conditions
         if iter_num > max_iters:
             if log_emissions:
-                emissions: float = tracker.stop()
```
Comment (Smith42) [inline scripts/train.py:712]: revert this

File: scripts/train.py [hunk 2]
```diff
@@ -643,73 +538,47 @@ def save_checkpoint(filename):
...
-                )
-                running_mfu = (
-                    mfu if running_mfu == -1.0 else 0.9 * running_mfu + 0.1 * mfu
-                )
+            if local_iter_num >= 5:
+                mfu = raw_model.estimate_mfu(batch_size * gradient_accumulation_steps, dt)
+                running_mfu = mfu if running_mfu == -1.0 else 0.9 * running_mfu + 0.1 * mfu
             if log_via_wandb:
                 wandb.log({"loss": lossf, "time": dt}, step=iter_num)
-            if log_emissions:
```
Comment (Smith42) [inline scripts/train.py:695]: Why are we removing the emissions logging?

File: config/pythia-like/ar_affine_001M.py
Comment (Smith42) [inline config/pythia-like/ar_affine_001M.py:1]: we need a better subdir than "study". "pythia-like"?

File: scripts/train.py [hunk 3]
```diff
@@ -164,29 +141,40 @@ def process_galaxy_wrapper(galdict, func):
         for k, v in globals().items()
         if not k.startswith("_") and isinstance(v, (int, float, bool, str))
     ]
-    exec(
-        open("src/astropt/configurator.py").read()
-    )  # overrides from command line or config file
-    config = {k: globals()[k] for k in config_keys}  # will be useful for logging
+    exec(open("src/astropt/configurator.py").read())  # overrides from CLI / config file
+    # MAE needs bidirectional attention; upgrade silently if left as the AR default.
```
Comment (Smith42) [inline scripts/train.py]: this shouldn't be silent

**Comment by ksd3** [comment]
@Smith42 #67 is updated and ready for another look:

- **Rebased onto current `main`** now that #63–66 are merged, so the conflict is gone — it shows as mergeable.
- It's now just the **unification delta**: one `scripts/train.py` handling both AR and MAE via the `objective` flag (sharding / checkpointing / tokeniser already live in `main`), `train_mae.py` removed, plus the 12 scaling-study configs in `config/study/`.
- Preserved all of the existing `train.py` comments + the `loss.png` / wandb image logging, and fixed the log checkpoint schedule to always pin the first **and** last step.
- Smoke-tested end-to-end on GPU for both objectives (train, eval, reconstruction previews, checkpoints, `load_astropt` reload).

It's still a draft — happy to mark it ready, or to split it differently if you'd prefer.

**Review by Smith42 — CHANGES_REQUESTED** [review]

File: scripts/train.py
Comment (ksd3) [reply scripts/train.py:1]: Did a manual check. Some of the comments that got deleted are because they moved to config instead

File: scripts/train.py (hunk 1 above)
Comment (ksd3) [reply scripts/train.py:712]: Reverted

File: scripts/train.py (hunk 2 above)
Comment (ksd3) [reply scripts/train.py:695]: Reverted

File: config/pythia-like/ar_affine_001M.py
Comment (ksd3) [reply config/pythia-like/ar_affine_001M.py:1]: Done

File: scripts/train.py (hunk 3 above)
Comment (ksd3) [reply scripts/train.py]: Fixed

**Review by Smith42 — APPROVED** [review]
Looks good to me

I have not tested this however, so there may be bugs we can fix in main before a versioned release.
